File: ios/Classes/RCacheFlutterHandler.swift
//
//  RCacheFlutterHandler.swift
//  rcache
//
//  Created by Rahmat Trinanda Pramudya Amar on 24/08/24.
//

import Flutter
import RCache

internal class RCacheFlutterHandler {
    
    internal class func common(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        if (call.method == RCacheFlutterMethod.Key.clear.rawValue) {
            RCache.common.clear()
            result("Success clear")
            return
        }
        
        guard let args = call.arguments as? Dictionary<String, Any> else {
            result(FlutterMethodNotImplemented)
            return
        }
        
        guard let key = args["key"] as? String else {
            result(FlutterError(code: "2", message: "RCache Key not defined", details: nil))
            return
        }
        
        switch call.method {
        case RCacheFlutterMethod.Key.Save.data.rawValue:
            if let data = args["value"] as? FlutterStandardTypedData {
                RCache.common.save(data: data.data, key: RCache.Key(key))
                result("Success saving")
            } else {
                result(FlutterError(code: "3", message: "Invalid format", details: nil))
            }
        case RCacheFlutterMethod.Key.Save.string.rawValue:
            RCache.common.save(string: args["value"] as! String, key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Save.bool.rawValue:
            RCache.common.save(bool: args["value"] as! Bool, key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Save.integer.rawValue:
            RCache.common.save(integer: args["value"] as! Int, key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Save.array.rawValue:
            RCache.common.save(array: args["value"] as! [Any], key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Save.dictionary.rawValue:
            RCache.common.save(dictionary: args["value"] as! [String:Any], key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Save.double.rawValue:
            RCache.common.save(double: args["value"] as! Double, key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Read.data.rawValue:
            result(RCache.common.readData(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.string.rawValue:
            result(RCache.common.readString(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.bool.rawValue:
            result(RCache.common.readBool(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.integer.rawValue:
            result(RCache.common.readInteger(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.array.rawValue:
            result(RCache.common.readArray(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.dictionary.rawValue:
            result(RCache.common.readDictionary(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.double.rawValue:
            result(RCache.common.readDouble(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.double.rawValue:
            RCache.common.remove(key: RCache.Key(key))
            result("Success saving")
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    internal class func credentials(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        if (call.method == RCacheFlutterMethod.Key.clear.rawValue) {
            RCache.credentials.clear()
            result("Success clear")
            return
        }
        
        guard let args = call.arguments as? Dictionary<String, Any> else {
            result(FlutterMethodNotImplemented)
            return
        }
        
        guard let key = args["key"] as? String else {
            result(FlutterError(code: "2", message: "RCache Key not defined", details: nil))
            return
        }
        
        switch call.method {
        case RCacheFlutterMethod.Key.Save.data.rawValue:
            if let data = args["value"] as? FlutterStandardTypedData {
                RCache.credentials.save(data: data.data, key: RCache.Key(key))
                result("Success saving")
            } else {
                result(FlutterError(code: "3", message: "Invalid format", details: nil))
            }
        case RCacheFlutterMethod.Key.Save.string.rawValue:
            RCache.credentials.save(string: args["value"] as! String, key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Save.bool.rawValue:
            RCache.credentials.save(bool: args["value"] as! Bool, key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Save.integer.rawValue:
            RCache.credentials.save(integer: args["value"] as! Int, key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Save.array.rawValue:
            RCache.credentials.save(array: args["value"] as! [Any], key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Save.dictionary.rawValue:
            RCache.credentials.save(dictionary: args["value"] as! [String:Any], key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Save.double.rawValue:
            RCache.credentials.save(double: args["value"] as! Double, key: RCache.Key(key))
            result("Success saving")
        case RCacheFlutterMethod.Key.Read.data.rawValue:
            result(RCache.credentials.readData(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.string.rawValue:
            result(RCache.credentials.readString(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.bool.rawValue:
            result(RCache.credentials.readBool(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.integer.rawValue:
            result(RCache.credentials.readInteger(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.array.rawValue:
            result(RCache.credentials.readArray(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.dictionary.rawValue:
            result(RCache.credentials.readDictionary(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.double.rawValue:
            result(RCache.credentials.readDouble(key: RCache.Key(key)))
        case RCacheFlutterMethod.Key.Read.double.rawValue:
            RCache.credentials.remove(key: RCache.Key(key))
            result("Success saving")
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    internal class func clear(result: @escaping FlutterResult) {
        RCache.clear()
        result("Success clear")
    }
}

File: android/src/main/kotlin/id/nesd/rcache_flutter/RCacheFlutterHandler.kt
package id.nesd.rcache_flutter

import id.nesd.rcache.RCache
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

internal object RCacheFlutterHandler {

    fun common(call: MethodCall, result: Result) {

        if (call.method == RCacheFlutterMethod.Key.clear) {
            RCache.common.clear()
            result.success("Success clear")
            return
        }

        val key = call.argument<String?>("key")

        if (key == null) {
            result.error("2", "RCache Key not defined", null)
            return
        }

        when (call.method) {
            RCacheFlutterMethod.Key.Save.uint8List -> {
                val byteArray: ByteArray? = call.argument<ByteArray>("value")
                if (byteArray == null) {
                    result.error("3", "Invalid format", null)
                } else {
                    RCache.common.save(byteArray, RCache.Key(key))
                    result.success("Success saving")
                }
            }

            RCacheFlutterMethod.Key.Save.string -> {
                RCache.common.save(call.argument<String>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Save.bool -> {
                RCache.common.save(call.argument<Boolean>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Save.integer -> {
                RCache.common.save(call.argument<Int>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Save.array -> {
                RCache.common.save(call.argument<List<Any>>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Save.map -> {
                RCache.common.save(call.argument<Map<String, Any>>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Save.double -> {
                RCache.common.save(call.argument<Double>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Read.uint8List -> {
                result.success(RCache.common.readByteArray(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.string -> {
                result.success(RCache.common.readString(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.bool -> {
                result.success(RCache.common.readBool(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.integer -> {
                result.success(RCache.common.readInteger(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.array -> {
                result.success(RCache.common.readArray<Any>(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.map -> {
                result.success(RCache.common.readMap<Any>(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.double -> {
                result.success(RCache.common.readDouble(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.remove -> {
                RCache.common.remove(RCache.Key(key))
                result.success("Success remove")
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    fun credentials(call: MethodCall, result: Result) {

        if (call.method == RCacheFlutterMethod.Key.clear) {
            RCache.credentials.clear()
            result.success("Success clear")
            return
        }

        val key = call.argument<String?>("key")

        if (key == null) {
            result.error("2", "RCache Key not defined", null)
            return
        }

        when (call.method) {
            RCacheFlutterMethod.Key.Save.uint8List -> {
                val byteArray: ByteArray? = call.argument<ByteArray>("value")
                if (byteArray == null) {
                    result.error("3", "Invalid format", null)
                } else {
                    RCache.credentials.save(byteArray, RCache.Key(key))
                    result.success("Success saving")
                }
            }

            RCacheFlutterMethod.Key.Save.string -> {
                RCache.credentials.save(call.argument<String>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Save.bool -> {
                RCache.credentials.save(call.argument<Boolean>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Save.integer -> {
                RCache.credentials.save(call.argument<Int>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Save.array -> {
                RCache.credentials.save(call.argument<List<Any>>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Save.map -> {
                RCache.credentials.save(call.argument<Map<String, Any>>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Save.double -> {
                RCache.credentials.save(call.argument<Double>("value")!!, RCache.Key(key))
                result.success("Success saving")
            }

            RCacheFlutterMethod.Key.Read.uint8List -> {
                result.success(RCache.credentials.readByteArray(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.string -> {
                result.success(RCache.credentials.readString(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.bool -> {
                result.success(RCache.credentials.readBool(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.integer -> {
                result.success(RCache.credentials.readInteger(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.array -> {
                result.success(RCache.credentials.readArray<Any>(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.map -> {
                result.success(RCache.credentials.readMap<Any>(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.Read.double -> {
                result.success(RCache.credentials.readDouble(RCache.Key(key)))
            }

            RCacheFlutterMethod.Key.remove -> {
                RCache.credentials.remove(RCache.Key(key))
                result.success("Success remove")
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    fun clear(result: Result) {
        RCache.clear()
        result.success("Success clear")
    }
}